[1.21] Rich/Component translations

BasiqueEvangelist opened the pull request:
This is effectively a port of [oωo-lib rich translations](https://docs.wispforest.io/owo/rich-translations/) (which was originally contributed to oωo by me) to NeoForge patches.

Currently marked as draft to get feedback on my patches. ~~Support for index text components isn't added yet.~~

BasiqueEvangelist committed:
add basic support for rich translations

index text components aren't added yet

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=1134) <br/>All committers have signed the CLA.

BasiqueEvangelist committed:
add inserting text content

patches are a bit cringe but meh

Fuzss commented on patches/net/minecraft/locale/Language.java.patch:
```diff
@@ -1,34 +1,74 @@
+         }
+     }
+ 
+-    public static void loadFromJson(InputStream p_128109_, BiConsumer<String, String> p_128110_) {
```
Please keep the vanilla method as an overload for multiloader projects.

BasiqueEvangelist replied:
I thought I already did?

BasiqueEvangelist committed:
keep original overloads of Language methods

BasiqueEvangelist committed:
keep original overloads of ClientLanguage methods

not really sure this is needed since they're all private, but might as well

Matyrobbrt commented on patches/net/minecraft/client/resources/language/ClientLanguage.java.patch:
```diff
@@ -1,14 +1,60 @@
++
++    private ClientLanguage(Map<String, String> p_118914_, Map<String, net.minecraft.network.chat.Component> componentStorage, boolean p_118915_) {
+         this.storage = p_118914_;
++        this.componentStorage = componentStorage;
```
This is a weird place to put the patch, maybe move it at the end?

BasiqueEvangelist replied:
Huh? Should I move the added constructor to the end or the field set?

Matyrobbrt replied:
The field set and the parameter in the added constructor.

BasiqueEvangelist replied:
I think I did it?

Matyrobbrt commented on patches/net/minecraft/locale/Language.java.patch:
```diff
@@ -1,34 +1,81 @@
+ 
+         for (Entry<String, JsonElement> entry : jsonobject.entrySet()) {
++            if (entry.getValue().isJsonArray()) {
++                var component = net.minecraft.network.chat.ComponentSerialization.CODEC.parse(com.mojang.serialization.JsonOps.INSTANCE, entry.getValue()).getOrThrow(com.google.gson.JsonParseException::new);
```
I would like to see a better error message (like the translation key and file it's parsing)

BasiqueEvangelist replied:
True. I'll probably also make this a logged error instead of an exception.

BasiqueEvangelist replied:
Added info about the translation key, the file it's parsing isn't passed in and will be logged anyways

Matyrobbrt commented on patches/net/minecraft/network/chat/contents/TranslatableContents.java.patch:
```diff
@@ -13,4 +13,77 @@
+     public <T> Optional<T> visit(FormattedText.StyledContentConsumer<T> p_237521_, Style p_237522_) {
+         this.decompose();
+ 
+-        for (FormattedText formattedtext : this.decomposedParts) {
```
If you don't indent this block, you don't need this patch.

BasiqueEvangelist replied:
That sounds like it'd make the resulting patched code less readable, but sure :+1:

Matyrobbrt commented on patches/net/minecraft/network/chat/contents/TranslatableContents.java.patch:
```diff
@@ -13,4 +13,77 @@
+     public <T> Optional<T> visit(FormattedText.ContentConsumer<T> p_237519_) {
+         this.decompose();
+ 
+-        for (FormattedText formattedtext : this.decomposedParts) {
```
Same as above.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/util/InsertingContents.java:
```diff
@@ -0,0 +1,83 @@
+
+    public static final ComponentContents.Type<InsertingContents> TYPE = new ComponentContents.Type<>(CODEC, "inserting");
+
+    private static final ThreadLocal<Deque<TranslatableContents>> translationStack = ThreadLocal.withInitial(ArrayDeque::new);
```
Should be named `TRANSLATION_STACK`.

Matyrobbrt committed:
Merge branch '1.21.x' into rich-translations

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/util/InsertingContents.java:
```diff
@@ -0,0 +1,83 @@
+            ExtraCodecs.NON_NEGATIVE_INT.fieldOf("index").forGetter(InsertingContents::index))
+            .apply(i, InsertingContents::new));
+
+    public static final ComponentContents.Type<InsertingContents> TYPE = new ComponentContents.Type<>(CODEC, "inserting");
```
The name of the component should ideally be more indicative of the fact that it's added by neo.

Matyrobbrt reviewed (changes requested):
Does the new component pose any problem for neo<->vanilla connections?

BasiqueEvangelist commented:
> Does the new component pose any problem for neo<->vanilla connections?

I think it will error at runtime on the vanilla end if the component is sent over the wire, but you aren't really supposed to use it outside of translations anyways.

marchermans commented:
> > Does the new component pose any problem for neo<->vanilla connections?
> 
> I think it will error at runtime on the vanilla end if the component is sent over the wire, but you aren't really supposed to use it outside of translations anyways.

Given that we forge translations will and do show up on vanilla clients, this seems like a blocker for me.

I would rather this system works purely as a parser and nothing else. In other words: the translatable component is aware of it innately, but when written to the wire the component is transformed to something vanilla can understand.

BasiqueEvangelist commented:
> > > Does the new component pose any problem for neo<->vanilla connections?
> > 
> > I think it will error at runtime on the vanilla end if the component is sent over the wire, but you aren't really supposed to use it outside of translations anyways.
> 
> Given that we forge translations will and do show up on vanilla clients, this seems like a blocker for me.
> 
> I would rather this system works purely as a parser and nothing else. In other words: the translatable component is aware of it innately, but when written to the wire the component is transformed to something vanilla can understand.

I'm saying that this will only explode if you actually instantiate an InsertingContents and send it. The wire format of translatable components isn't changed. 
Ideally the best solution would be to disallow sending inserting components directly to clients, since they're only supposed to be used in lang files anyways

marchermans commented:
> thing vanilla can understand.

Then maybe encapsulate this somehow, using for example package private access or sealed classes.

BasiqueEvangelist committed:
should™ address reviews

BasiqueEvangelist commented:
> Then maybe encapsulate this somehow, using for example package private access or sealed classes.

I don't know what you mean by that (InsertingContents needs to be public since it's referenced by an MC class), but I slapped an `@ApiStatus.Internal` on it

BasiqueEvangelist committed:
add tests for rich translations

this fails on dedicated servers right now as rich translations aren't
fully loaded there

BasiqueEvangelist committed:
add dedicated server support for rich translations

dedicated server tests should work now

HenryLoenwind commented:
How would this work with CrowdIn translations?

Also, aside from the link in this PR there is zero documentation. Not only is the availability of this undiscoverable for modders, finding out how to use it is a chore.

In my opinion, mixing the translation with the formatting is the wrong approach. Formatting needs to be *applied* to translations, the same way placeholders are applied. Having colours and styling isn't a property of the translation; it's a property of the usage of those translations. The translation strings need a way to markup which parts of the string are what and the code using it needs to supply the formatting needed for the place the string is used.

BasiqueEvangelist commented:
> How would this work with CrowdIn translations?

I honestly don't really know, though CrowdIn would probably prefer a more string-like format like MiniMessage.

> Also, aside from the link in this PR there is zero documentation. Not only is the availability of this undiscoverable for modders, finding out how to use it is a chore.

Do you think the PR itself should've added documentation anywhere? Where to? If you're talking about https://docs.neoforged.net/, I'll get to that soon™ (maybe).

> In my opinion, mixing the translation with the formatting is the wrong approach. Formatting needs to be _applied_ to translations, the same way placeholders are applied.

How do you envision that working? I sort of agree with that (although I think that users being able to customize message formatting is nice, and I much prefer putting formatting in the language files so that it doesn't mess with code).

HenryLoenwind commented:
My guess is that CrowdIn would either reject the whole json file or ignore those entries. A good way to find out would have been to convert at least one of the translations in neo's lang file, even if the style was just "none".

> Do you think the PR itself should've added documentation anywhere? Where to?

```java
/**
  * This is javadoc. It goes on methods and classes to explain what they are there for.
  **/
```

That would be a good start, especially for the class you added. There also are lots of patches, most of them so esoteric that I don't have any clue what they do, even after reading the code at least ten times today.

> How do you envision that working?

Semantic markup.

Just a random way to do it would be `"'%0$m{%1$s}' will be lost forever! %2$m{(A long time!)}"`, and having the Style objects as normal parameters in the list. Alternatively, they could be in a second list that's given as an independent parameter---more API changes and more patches, but that would allow a modder to use a single static format list and use it everywhere, leaving the choice of which styles to use to the translator (although that's not the job a translator is there for). Using %m has the advantage that it's an illegal placeholder for vanilla, which means there's next to no chance anyone would use it accidentally.

Other ways, like xml-style named markup (`"'<name>%s</name>' will be lost forever! <warning>(A long time"</warning>"`) have more potential for conflicts. MessageFormat-like markup (`"'{name:%s}' will be lost forever! {warning:(A long time"}"`) would also be a possibility; higher potential for conflicts, but {} are rather rare in translations and by checking the "{"+tag+":" format, it would become even less likely. Those also require the Styles to be put in as Maps, i.e. forcing an API change for creating translatable Components.

I also would be concerned about mixing code an presentation---but we don't have a presentation layer in Minecraft. All the UI stuff, down to the positioning in pixels, is handled in the code. Having a proper presentation layer that is decoupled from the code would be nice, and then that part certainly would go there, but that's wishful thinking.

---

PS: And the reason why I read that code so often is that I wanted to apply the pluralisation patches #1269 to it, too. However, I still have no clue where or how.

HenryLoenwind commented:
PS: If I sound grumpy that's because I'm in pain. Got a nerve channel in my right shoulder that likes to flare up every couple of years. :(
